networkx/algorithms/centrality/betweenness.py

```diff
@@ -394,3 +397,39 @@ def _rescale_e(betweenness, n, normalized, directed=False, k=None):
         for v in betweenness:
             betweenness[v] *= scale
     return betweenness
+
+
+@not_implemented_for("graph")
+def _add_edge_keys(G, betweenness, weight=None):
+    r"""Adds the corrected betweenness centrality (BC) values for multigraphs.
+
+    Parameters
+    ----------
+    G : NetworkX graph.
+
+    betweenness : dictionary
+        Dictionary mapping adjacent node tuples to betweenness centrality values.
+
+    weight : string or function
+        See `_get_edge_keys` for details. Defaults to `None`.
+
+    Returns
+    -------
+    edges : dictionary
+        The parameter `betweenness` including edges with keys and their
+        betweenness centrality values.
+
+    The BC value is divided among edges of equal weight.
+    """
+    _weight = _weight_function(G, weight)
+
+    edge_bc = dict.fromkeys(G.edges, 0.0)
+    for u, v in betweenness:
+        d = G[u][v]
+        wt = _weight(u, v, d)
+        keys = [k for k in d if d[k].get(weight, 1) == wt]
```

Comment:
Are we being consistent here in how we handle the `weight` parameter?  Are we going to allow it to be "a function or a string" or just "a string".  I think the current code works if it is a string, but doesn't work if `weight` is a function.  That's because the `d[k].get(weight, 1)` won't work correctly if `weight` is a function. 

If we restrict `weight` to be a string, we should probably avoid using `_weight_function` and just use a loop over all edges between the node-pairs, computing the minimal weight *and* the number of edges with that weight.  But it'd be nice to have this all work with a `weight` function too.  I'll think about how to do that and am open to suggestions.

Reply:
I think we should build our own _weight_function here and prepare for using a more flexible `_weight_function` type decorator like the one discussed in #4085 and #4114.  Something like this:
```python
If not callable(weight):
    _weight = weight
    def weight(u, v, key_dict):
        min_wt, keys = float(‘inf’), [] 
        for k, data_dict in key_dict.items():
            wt = data_dict.get(_weight, 1)
            if wt < min_wt:
                min_wt, keys = wt, [k]
            elif wt == min_wt:
                keys.append(k)
        return min_wt, keys 
```

Reply:
Good catch! I think having a global solution definitely favors maintainability. However, I don't think until then our solution should be polymorphic in the return type. If a user hands us a callable weight we'd fail trying to query the keys.

An interim solution would be to create a dict with just the edge value:

~~~
keys = [k for k in d if _weight(u, v, {k: d[k]}) == wt]
~~~

This minimizes code duplication and is consistent with the usage of current callable weights. I've tested it already.

If other parts of networkx also need to be revisited regarding how they act on multigraphs, I don't think inlining getting those keys has such a big impact on performance. So this solution could be reused without having to impose changes to any callable weights.
That said, I think having a polymorphic return type like proposed in the global solution is highly advisable. This would enforce authors to think about what cases with parallel edges of equal weight.

Reply:
Nice trick for making the weight function treat a multiedge like it is the only edge between those nodes. :}